theano/compile/builders.py

```diff
@@ -115,40 +173,197 @@ def __hash__(self):
+                        out.zeros_like().astype(theano.config.floatX))
+                else:
+                    all_rops_l.append(rov(self.local_inputs, eval_points))
+        elif self._rop_op is undef:
+            all_rops_l = [
+                out.zeros_like().astype(theano.config.floatX)
+                for out in self.local_outputs]
+        else:
+            all_rops_l = self._rop_op(self.local_inputs, eval_points)
+            if not isinstance(all_rops_l, (tuple, list)):
+                all_rops_l = [all_rops_l]
+            if len(all_rops_l) != len(self.local_outputs):
+                raise ValueError(
+                    'Rop overriding function %s should return list of '
+                    '%d outputs, got %d' % (
+                        self._rop_op,
+                        len(self.local_outputs),
+                        len(all_rops_l)),
+                    self._rop_op)
+        self._rop_op = type(self)(
+            inputs=self.local_inputs + eval_points,
+            outputs=all_rops_l,
+            inline=self.is_inline, on_unused_input='ignore')
+        self._rop_op_is_cached = True
+
+    def get_grad_op(self):
+        """
+        getter method for self._grad_op
+        """
+        if not self._grad_op_is_cached:
+            self._recompute_grad_op()
+        return self._grad_op
 
-    def prepare_node(self, node, storage_map, compute_map, impl):
-        if not hasattr(self, "fn") and impl == 'py':
-            self.fn = orig_function(self.new_inputs,
-                                    self.new_outputs,
-                                    **self.kwargs)
+    def get_rop_op(self):
+        """
+        getter method for self._rop_op
+        """
+        if not self._rop_op_is_cached:
+            self._recompute_rop_op()
+        return self._rop_op
 
-    def perform(self, node, inputs, outputs):
-        variables = self.fn(*inputs)
-        assert len(variables) == len(outputs)
-        for output, variable in zip(outputs, variables):
-            # TODO: when function's output-borrowing semantics are correct,
-            # we wont need this copy anymore
-            output[0] = variable.copy()
+    def set_rop_overrides(self, rop_overrides):
+        """
+        Set R_op overrides, see help(theano.OpFromGraph) for syntax
+        This will completely remove any previously set R_op overrides
+
+        """
+        self._rop_op = rop_overrides
+        self._rop_op_is_cached = False
+
+    def set_grad_overrides(self, grad_overrides):
+        """
+        Set gradient overrides, see help(theano.OpFromGraph) for syntax
+        This will completely remove any previously set gradient overrides
+
+        """
+        self._grad_op = grad_overrides
+        self._grad_op_is_cached = False
+
+    def R_op(self, inputs, eval_points):
+        if not self._rop_op_is_cached:
+            self._recompute_rop_op()
+        return self._rop_op(*(list(inputs) + list(eval_points)), return_list=True)
```

Comment:
`_recompute_rop_op` will convert whatever override user supplied to an OfG instance, so no need to worry that.